contracts/origination/OriginationControllerRefinance.sol

```diff
@@ -0,0 +1,203 @@
+        if (_originationConfig == address(0)) revert OCR_ZeroAddress("_originationConfig");
+        if (_loanCore == address(0)) revert OCR_ZeroAddress("_loanCore");
+        if (_feeController == address(0)) revert OCR_ZeroAddress("_feeController");
+
+        originationConfig = IOriginationConfiguration(_originationConfig);
+        loanCore = ILoanCore(_loanCore);
+        feeController = IFeeController(_feeController);
+    }
+
+    /**
+     * @notice Refinances an active loan. This function can only be called by a new lender account.
+     *         There is no signature required from the borrower. The new loan terms will be validated
+     *         and the new loan will be created. The old loan will be closed and replaced by the new
+     *         one. The new lender will pay the old lender the full repayment amount.
+     *
+     * @param loanId                             The ID of the loan to be refinanced.
+     * @param newTerms                           The new loan terms.
+     *
+     * @return newLoanId                         The ID of the new loan.
+     */
+    function refinanceLoan(
+        uint256 loanId,
+        LoanLibrary.LoanTerms calldata newTerms
+    ) external override returns (uint256 newLoanId) {
+        LoanLibrary.LoanData memory data = loanCore.getLoan(loanId);
+
+        // validate refinance
+        _validateRefinance(data, newTerms);
+
+        address borrower = IERC721(loanCore.borrowerNote()).ownerOf(loanId);
+
+        newLoanId = _refinance(
+            loanId,
+            newTerms,
+            borrower,
+            msg.sender
+        );
+    }
+
+    /**
+     * @notice Validates the new loan terms for a refinanced loan. The new APR must be at least 5%
+     *         lower than the old APR. The new principal amount cannot be larger. The new due date
+     *         cannot be shorter than the old due date and the collateral and payable currency must
+     *         be the same.
+     *
+     * @param oldLoanData                        The loan data of the loan being refinanced.
+     * @param newTerms                           The new loan terms.
+     */
+    // solhint-disable-next-line code-complexity
+    function _validateRefinance(
+        LoanLibrary.LoanData memory oldLoanData,
+        LoanLibrary.LoanTerms calldata newTerms
+    ) internal view {
+        // cannot refinance a loan that has already been repaid
+        if (oldLoanData.state != LoanLibrary.LoanState.Active) revert OCR_InvalidState(oldLoanData.state);
+
+        // cannot refinance a loan before it has been active for 2 days
+        if (block.timestamp < oldLoanData.startDate + 2 days) revert OCR_TooEarly(oldLoanData.startDate + 2 days);
+
+        // new interest rate APR must be lower than old interest rate by minimum
+        uint256 aprMinimumScaled = oldLoanData.terms.interestRate * Constants.BASIS_POINTS_DENOMINATOR -
+            (oldLoanData.terms.interestRate * MINIMUM_INTEREST_CHANGE);
```

Comment:
How come we don't need the `Constants.BASIS_POINTS_DENOMINATOR` in the second term of the subtraction, where we also read `terms.interestRate`? Does `MINIMUM_INTEREST_CHANGE` include the scaling factor?

Reply:
Yeah the `MINIMUM_INTEREST_CHANGE` value has this included, so 5% would be 10,000/20 = 500bps, 10% 10,000/10 = 1000 bps